Pull Request to add examples/webgl_loader_viewer.html - Model viewer

jsidd opened the pull request:
This addition is based on this mrdoob/three.js issue:
[https://github.com/mrdoob/three.js/issues/5149](https://github.com/mrdoob/three.js/issues/5149)

While #5149 does seem to be an old issue/request and there are current many loader examples (at least one for each loader), it might still be useful to have an model viewer that could open ultimately all the supported model types. 

As such examples/wegl_loader_viewer.html is my initial attempt at meeting this goal. 

Currently my viewer can load most Mesh based model format but not Geometry based models. 

The camera currently zooms in on a bounding box but it does not work for all models. 

The controls are currently set on OrbitControls and not TrackballControls because I wanted to add a simple html dropdown to select a model. OrbitControls does support zoom in/out (scroll), rotate (primary click and drag), pan (alternate click and drag).  It might be possible to use TrackballControls with a more complex ui (such as settings on other examples). 

*****
Issue/Request:
[https://github.com/mrdoob/three.js/issues/5149](https://github.com/mrdoob/three.js/issues/5149)

Model viewer example #5149

"This is a request to add a model viewer example.
I think it's a very common need to display a model, pan, rotate and zoom.

Features:

Load models of different formats with plugable loaders. 
It can help unify the loaders API if needed.
Auto zoom the model to a reasonable initial value. Allow zoom in/out.
Use TrackballControls to rotate the model.
Allow pan."

jsidd committed:
examples/webgl_loader_viewer.html - Initial Commit

supports Babylon, obj, awd, dae, fbx, gltf, json
load a model from threejs local/website via dropdown
load a custom model by inputing local or web url

jsidd committed:
added assimp json support, pcd points support, sea, and vrml (ver 2?) support

jsidd committed:
examples/wegl_loader_viewer.html additions

resetCamera() attempts to find best view for loaded model
OrbitControls are reset every time new model is loaded ( reloadModel() )
post loader fixes for files for which resetCamera() did not work optimally
pcd is supported but has issues (view and orientation of model)
dae is supported but loader.traverse() fails on last child?
added debug print toggle option - display div with debug info - displays info about loaded model nodes
(note - some node types of loaded model are not added to the scene: bones and lights for example)

Current to do list:
1//to do: improvement - resetCamera() does not always result in best view
2//to do: improvement - setupLights() improve to use better lights
3//to do: improvement - loader.load() note: sea loader does not use onload function - loader.load 2nd param
4//to do: improvement - post loader.load() fixes - handle camera for these files automatically
5//to do: improvement - Zaghetto.pcd need to flip pcd model upside down as needed
6//to do: improvement - Zaghetto.pcd find correct camera settings
7//to do: bug - fileScene.traverse() dae error on traversing last child?
8//to do: improvement - fileScene.traverse() handle nested groups?
9//to do: bug - with gltf file can not move model unless material is changed
A//to do: improvement - selectLoader()/OnLoadScene() add support for Geometry Loaders
//right now only supports Mesh Loaders

jsidd committed:
Model Viewer addition, original issue #5149 - Fixed equality operator to use strict equality per Mr.doob's Code Style™

mrdoob commented:
For reference, this is the output:
https://rawgit.com/jsidd/three.js/jsidd/examples/webgl_loader_viewer.html

bchevalier commented:
You mention that "the camera currently zooms in on a bounding box but it does not work for all models".  Is it because there is an issue with the ```new THREE.Box3().setFromObject(scene)``` method?

bchevalier commented on examples/webgl_loader_viewer.html:
```diff
@@ -0,0 +1,410 @@
+		setupLights();
+
+		//to do: improvement - loader.load() note: sea loader does not use onload function - loader.load 2nd param
+		loader.load( objPath , function ( myScene ) {
```
Could there be a slightly simpler API to load any model?

Also, is it convenient for anyone willing to use this API to have to copy/past this code? (which they might have a hard-time finding)

bchevalier commented on examples/webgl_loader_viewer.html:
```diff
@@ -0,0 +1,410 @@
+		//to do: improvement - post loader.load() fixes - handle camera for these files automaticaly
+		if (fileName==="mascot.tjs.sea")
+		{
+			camera.position.z = 2000;
```
does the need for this offset came from the bounding box calculation issue?
I am not sure it is easy to understand why this is necessary.

jsidd replied:
This '.sea' model seems to ignore bounding box calculation; still working on it.

bchevalier replied:
nothing related to the ```.size()``` method (instead of ```.getSize()```)?

jsidd replied:
I changed it and it had not effect. If I don't move the camera with this line, the camera is actually inside the model.

jsidd replied:
I fixed the camera of .tjs.sea model by using a custom param with the SEA loader, as SEA has its own node adding to scene mechanism and seems to ignore the camera settings otherwise.

bchevalier commented on examples/webgl_loader_viewer.html:
```diff
@@ -0,0 +1,410 @@
+
+				if (objName === "skull.babylon" || objName === "duck.gltf" )
+				{
+					//to do: bug - with gltf file can not move model unless material is changed
```
I wonder, is it a know bug?

jsidd replied:
I am not sure yet, still looking at it. In my 5th commit, I left node materials as is.

jsidd replied:
Finally fixed gltf camera, controls issue, apparently this is required: `THREE.GLTFLoader.Shaders.update(scene, camera)`
as shown below:

```
function animate() {

requestAnimationFrame( animate );
if (fileType === "gltf")
{
THREE.GLTFLoader.Shaders.update(scene, camera);
}
render();

}
```

looeee commented on examples/webgl_loader_viewer.html:
```diff
@@ -0,0 +1,410 @@
+			//to do: bug - fileScene.traverse() dae error on traversing last child?
+
+			//to do: improvement - fileScene.traverse() handle nested groups?
+			if ( object.constructor.name === "Group" ) {
```
@jsidd this may cause problems with three.min.js

You should use object.type === "Group" instead.
See #10105

looeee replied:
@jsidd yes. Have a look at the console output of this [fiddle](http://jsfiddle.net/looeee/9hqkqxjj/)

object.constructor === THREE.Group on the other hand does work, in this case, but as far as I'm aware is generally not a reccomended pattern as constructors can be overwritten.

mrdoob replied:
Actually, it should be `object instanceof THREE.Group`.

jsidd replied:
@looeee @mrdoob Thank you.

jsidd commented on examples/webgl_loader_viewer.html:
```diff
@@ -0,0 +1,410 @@
+			<option value="models/assimp/jeep/jeep.assimp.json">models/assimp/jeep/jeep.assimp.json</option>
+			<option value="models/pcd/Zaghetto.pcd">models/pcd/Zaghetto.pcd</option>
+			<option value="models/sea3d/mascot.tjs.sea">models/sea3d/mascot.tjs.sea</option>
+			<option value="models/vrml/house.wrl">models/vrml/house.wrl</option>
```
I think the most compatible way would to use an index file of all models in three.js examples/models and examples/obj directories. This index file would need to be added to the git repo, automatically or systematically generated and maintained.

At first, I did think of getting file lists directly from these two directories, but it is not possible because of client side javascript security restriction. I did run across articles explaining that File API were considered for HTML5 specs that would allow to list content of directories and retrieve full local path but it seems to be only supported by Chrome and Opera at the moment: http://caniuse.com/#feat=filesystem

Also, if we used node.js with three.js,  we could use server side javascript and get file lists that way.  So a three.js model viewer version using node.js might be more flexible. It could be part of the npm three.js-node package.

I researched how we could build an index file, and found that with the help the following find command on Mac OS X, we can generate a list that can used to generate an index file: 

find . -type f -print0 | xargs -0 ls -t

For the "examples/models" directory, it currently displays:

./.DS_Store
./collada/.DS_Store
./animated/.DS_Store
./obj/.DS_Store
./skinned/marine/marine_anims_core.blend
./skinned/marine/marine_anims_core.json
./skinned/marine/marine_anims_single.blend
./skinned/marine/marine_anims_single.json
./skinned/marine/marine_ik.blend
./skinned/marine/marine_ik.json
./skinned/simple/simple.blend
./skinned/simple/simple.js
./stl/ascii/pr2_head_pan.stl
./stl/ascii/pr2_head_tilt.stl
./stl/ascii/slotted_disk.stl
./stl/binary/colored.stl
./stl/binary/pr2_head_pan.stl
./stl/binary/pr2_head_tilt.stl
./utf8/James_Body_Lores.jpg
./utf8/James_EyeLashBotTran.png
./utf8/James_EyeLashTopTran.png
./utf8/James_Eye_Green.jpg
./utf8/James_Eye_Inner_Green.jpg
./utf8/James_Face_Color_Hair_Lores.jpg
./utf8/James_Mouth_Gum_Lores.jpg
./utf8/James_Tongue_Lores.jpg
./utf8/MCasShoe1TEX_Lores.jpg
./utf8/MJeans1TEX_Lores.jpg
./utf8/MTshirt3TEX_Lores.jpg
./utf8/Nail_Hand_01_Lores.jpg
./utf8/WaltHi.js
./utf8/WaltHi.utf8
./utf8/ben.js
./utf8/ben.utf8
./utf8/ben_dds.js
./utf8/dds/James_Body_Lores.dds
./utf8/dds/James_EyeLashBotTran.dds
./utf8/dds/James_EyeLashTopTran.dds
./utf8/dds/James_Eye_Green.dds
./utf8/dds/James_Eye_Inner_Green.dds
./utf8/dds/James_Face_Color_Hair_Lores.dds
./utf8/dds/James_Mouth_Gum_Lores.dds
./utf8/dds/James_Tongue_Lores.dds
./utf8/dds/MCasShoe1TEX_Lores.dds
./utf8/dds/MJeans1TEX_Lores.dds
./utf8/dds/MTshirt3TEX_Lores.dds
./utf8/dds/Nail_Hand_01_Lores.dds
./utf8/hand.jpg
./utf8/hand.js
./utf8/hand.utf8
./vrml/house.wrl
./vrml/simple.wrl
./vtk/bunny.vtk
./vtk/cube_ascii.vtp
./vtk/cube_binary.vtp
./vtk/cube_no_compression.vtp
./vtk/liver.vtk
./obj/vive-controller/onepointfive_spec.png
./obj/vive-controller/onepointfive_texture.png
./obj/vive-controller/vr_controller_vive_1_5.obj
./pcd/Zaghetto.pcd
./platform/platform.jpg
./platform/platform.json
./ply/ascii/dolphins.ply
./ply/ascii/dolphins_colored.ply
./ply/binary/Lucy100k.ply
./ply/binary/dolphins_be.ply
./ply/binary/dolphins_le.ply
./sea3d/keyframe.sea
./sea3d/keyframe.tjs.sea
./sea3d/mascot.sea
./sea3d/mascot.tjs.sea
./sea3d/morph.sea
./sea3d/morph.tjs.sea
./sea3d/robot.o3dgc.tjs.sea
./sea3d/robot.sea
./sea3d/robot.tjs.sea
./sea3d/skin.sea
./sea3d/skin.tjs.sea
./sea3d/sound.sea
./sea3d/sound.tjs.sea
./skin_and_morph.dae
./skinned/UCS/skins/Asian_Male.jpg
./skinned/UCS/skins/Black_Female.jpg
./skinned/UCS/skins/Caucasion_Female.jpg
./skinned/UCS/skins/Caucasion_Male.jpg
./skinned/UCS/skins/Highlighted_Muscles.jpg
./skinned/UCS/skins/Indian_Male.jpg
./skinned/UCS/umich_ucs.js
./skinned/UCS_config.json
./skinned/knight.js
./skinned/marine/M4.png
./skinned/marine/MarineCv2_color.jpg
./skinned/marine/m4.blend
./skinned/marine/m4.js
./skinned/marine/marine_anims.json
./skinned/marine/marine_anims_all.blend
./skinned/marine/marine_anims_all.json
./3mf/cube_gears.3mf
./amf/rook.amf
./animated/flamingo.js
./animated/horse.js
./animated/monster/monster.blend
./animated/monster/monster.jpg
./animated/monster/monster.js
./animated/parrot.js
./animated/sittingBox.js
./animated/stork.js
./assimp/interior/image1.jpg
./assimp/interior/image2.jpg
./assimp/interior/interior.3ds
./assimp/interior/interior.assimp.json
./assimp/interior/wall.jpg
./assimp/jeep/jeep.assimp.json
./assimp/jeep/jeep1.jpg
./assimp/jeep/jeep1.ms3d
./assimp/jeep/jeep1.readme.txt
./awd/simple/simple.awd
./awd/simple/textures/texture.jpg
./babylon/skull.babylon
./bvh/pirouette.bvh
./collada/VWS_B_Male2-2.jpg
./collada/avatar.dae
./collada/kawada-hironx.dae
./collada/monster/monster.dae
./collada/monster/monster.jpg
./collada/monster/readme.txt
./collada/multimaterial.dae
./collada/pump/pump.dae
./collada/pump/pump_body.jpg
./collada/pump/pump_gears.jpg
./collada/pump/pump_metalreflect.jpg
./ctm/LeePerry.ctm
./ctm/WaltHead.ctm
./ctm/ben.ctm
./ctm/camaro/camaro.ctm
./ctm/camaro/camaro.js
./ctm/camaro/car-ao.png
./ctm/camaro/plane-ao-256.png
./ctm/hand.ctm
./fbx/Char_UV_Texture.gif
./fbx/xsi_man.fbx
./fbx/xsi_man_skinning.fbx
./gltf/CesiumMan/README.txt
./gltf/CesiumMan/glTF-MaterialsCommon/Cesium_Man.bin
./gltf/CesiumMan/glTF-MaterialsCommon/Cesium_Man.gltf
./gltf/CesiumMan/glTF-MaterialsCommon/Cesium_Man.jpg
./gltf/CesiumMan/glTF/Cesium_Man.bin
./gltf/CesiumMan/glTF/Cesium_Man.gltf
./gltf/CesiumMan/glTF/Cesium_Man.jpg
./gltf/CesiumMan/glTF/Cesium_Man0FS.glsl
./gltf/CesiumMan/glTF/Cesium_Man0VS.glsl
./gltf/CesiumMilkTruck/README.txt
./gltf/CesiumMilkTruck/glTF-MaterialsCommon/CesiumMilkTruck.bin
./gltf/CesiumMilkTruck/glTF-MaterialsCommon/CesiumMilkTruck.gltf
./gltf/CesiumMilkTruck/glTF-MaterialsCommon/CesiumMilkTruck.png
./gltf/CesiumMilkTruck/glTF/CesiumMilkTruck.bin
./gltf/CesiumMilkTruck/glTF/CesiumMilkTruck.gltf
./gltf/CesiumMilkTruck/glTF/CesiumMilkTruck.png
./gltf/CesiumMilkTruck/glTF/CesiumMilkTruck0FS.glsl
./gltf/CesiumMilkTruck/glTF/CesiumMilkTruck0VS.glsl
./gltf/CesiumMilkTruck/glTF/CesiumMilkTruck1FS.glsl
./gltf/CesiumMilkTruck/glTF/CesiumMilkTruck1VS.glsl
./gltf/duck/README.txt
./gltf/duck/glTF-MaterialsCommon/duck.bin
./gltf/duck/glTF-MaterialsCommon/duck.gltf
./gltf/duck/glTF-MaterialsCommon/duckCM.png
./gltf/duck/glTF/duck.bin
./gltf/duck/glTF/duck.gltf
./gltf/duck/glTF/duck0FS.glsl
./gltf/duck/glTF/duck0VS.glsl
./gltf/duck/glTF/duckCM.png
./gltf/monster/glTF-MaterialsCommon/monster.bin
./gltf/monster/glTF-MaterialsCommon/monster.gltf
./gltf/monster/glTF-MaterialsCommon/monster.jpg
./gltf/monster/glTF/monster.bin
./gltf/monster/glTF/monster.gltf
./gltf/monster/glTF/monster.jpg
./gltf/monster/glTF/monster0FS.glsl
./gltf/monster/glTF/monster0VS.glsl
./gltf/monster/readme.txt
./json/pressure.json
./json/scene-animation.json
./json/teapot-claraio.json
./md2/ogro/Ogro.txt
./md2/ogro/ogro.md2
./md2/ogro/skins/arboshak.png
./md2/ogro/skins/ctf_b.png
./md2/ogro/skins/ctf_r.png
./md2/ogro/skins/darkam.png
./md2/ogro/skins/freedom.png
./md2/ogro/skins/gib.png
./md2/ogro/skins/gordogh.png
./md2/ogro/skins/grok.jpg
./md2/ogro/skins/igdosh.png
./md2/ogro/skins/khorne.png
./md2/ogro/skins/nabogro.png
./md2/ogro/skins/ogrobase.png
./md2/ogro/skins/sharokh.png
./md2/ogro/skins/weapon.jpg
./md2/ogro/weapon.md2
./md2/ratamahatta/ratamahatta.md2
./md2/ratamahatta/skins/ctf_b.png
./md2/ratamahatta/skins/ctf_r.png
./md2/ratamahatta/skins/dead.png
./md2/ratamahatta/skins/gearwhore.png
./md2/ratamahatta/skins/ratamahatta.png
./md2/ratamahatta/skins/w_bfg.png
./md2/ratamahatta/skins/w_blaster.png
./md2/ratamahatta/skins/w_chaingun.png
./md2/ratamahatta/skins/w_glauncher.png
./md2/ratamahatta/skins/w_hyperblaster.png
./md2/ratamahatta/skins/w_machinegun.png
./md2/ratamahatta/skins/w_railgun.png
./md2/ratamahatta/skins/w_rlauncher.png
./md2/ratamahatta/skins/w_shotgun.png
./md2/ratamahatta/skins/w_sshotgun.png
./md2/ratamahatta/skins/weapon.png
./md2/ratamahatta/w_bfg.md2
./md2/ratamahatta/w_blaster.md2
./md2/ratamahatta/w_chaingun.md2
./md2/ratamahatta/w_glauncher.md2
./md2/ratamahatta/w_hyperblaster.md2
./md2/ratamahatta/w_machinegun.md2
./md2/ratamahatta/w_railgun.md2
./md2/ratamahatta/w_rlauncher.md2
./md2/ratamahatta/w_shotgun.md2
./md2/ratamahatta/w_sshotgun.md2
./md2/ratamahatta/weapon.md2
./mmd/LICENSE
./mmd/Readme.txt
./mmd/audios/readme.txt
./mmd/audios/wavefile_short.mp3
./mmd/miku/eyeM2.bmp
./mmd/miku/miku_v2.pmd
./mmd/miku/readme.txt
./mmd/miku/readme_miku_v2.txt
./mmd/vmds/readme.txt
./mmd/vmds/readme_wavefile.txt
./mmd/vmds/readme_wavefile_camera.txt
./mmd/vmds/wavefile_camera.vmd
./mmd/vmds/wavefile_v2.vmd
./mmd/vpds/01.vpd
./mmd/vpds/02.vpd
./mmd/vpds/03.vpd
./mmd/vpds/04.vpd
./mmd/vpds/05.vpd
./mmd/vpds/06.vpd
./mmd/vpds/07.vpd
./mmd/vpds/08.vpd
./mmd/vpds/09.vpd
./mmd/vpds/10.vpd
./mmd/vpds/11.vpd
./mmd/vpds/readme.txt
./molecules/.htaccess
./molecules/Al2O3.pdb
./molecules/aspirin.pdb
./molecules/buckyball.pdb
./molecules/caf2.pdb
./molecules/caffeine.pdb
./molecules/cholesterol.pdb
./molecules/cocaine.pdb
./molecules/cu.pdb
./molecules/cubane.pdb
./molecules/diamond.pdb
./molecules/ethanol.pdb
./molecules/glucose.pdb
./molecules/graphite.pdb
./molecules/lsd.pdb
./molecules/lycopene.pdb
./molecules/nacl.pdb
./molecules/nicotine.pdb
./molecules/ybco.pdb
./nrrd/I.nrrd
./obj/cerberus/Cerberus.obj
./obj/cerberus/Cerberus_A.jpg
./obj/cerberus/Cerberus_M.jpg
./obj/cerberus/Cerberus_N.jpg
./obj/cerberus/Cerberus_R.jpg
./obj/emerald.obj
./obj/tree.obj

For the "examples/obj" directory, it currently displays:

./.DS_Store
./walt/WaltHead_bin.bin
./walt/WaltHead_bin.js
./walt/WaltHead_slim.js
./Bird.js
./Qrcode.js
./Suzanne.js
./WaltHeadLo.js
./camaro/.htaccess
./camaro/CamaroNoUv_bin.bin
./camaro/CamaroNoUv_bin.js
./cubecolors/.htaccess
./cubecolors/cube_fvc.blend
./cubecolors/cube_fvc.js
./cubecolors/cubecolors.blend
./cubecolors/cubecolors.js
./f50/.htaccess
./f50/F50NoUv_bin.bin
./f50/F50NoUv_bin.js
./female02/.htaccess
./female02/01_-_Default1noCulling.JPG
./female02/02_-_Default1noCulling.JPG
./female02/03_-_Default1noCulling.JPG
./female02/Female02_bin.bin
./female02/Female02_bin.js
./female02/Female02_slim.js
./female02/female02.mtl
./female02/female02.obj
./female02/readme.txt
./gallardo/.htaccess
./gallardo/GallardoNoUv_bin.bin
./gallardo/GallardoNoUv_bin.js
./gallardo/parts/gallardo_body_bin.bin
./gallardo/parts/gallardo_body_bin.js
./gallardo/parts/gallardo_wheel_bin.bin
./gallardo/parts/gallardo_wheel_bin.js
./leeperrysmith/.htaccess
./leeperrysmith/Infinite-Level_02_Disp_NoSmoothUV-4096.jpg
./leeperrysmith/Infinite-Level_02_Tangent_SmoothUV.jpg
./leeperrysmith/LeePerrySmith.js
./leeperrysmith/LeePerrySmith.obj
./leeperrysmith/LeePerrySmith_License.txt
./leeperrysmith/Map-COL.jpg
./leeperrysmith/Map-SPEC.jpg
./lightmap/lightmap-ao-shadow.png
./lightmap/lightmap.js
./lightmap/readme.txt
./lightmap/rocks.jpg
./lightmap/stone.jpg
./lucy/.htaccess
./male02/.htaccess
./male02/01_-_Default1noCulling.JPG
./male02/01_-_Default1noCulling.dds
./male02/Male02_bin.bin
./male02/Male02_bin.js
./male02/Male02_dds.js
./male02/Male02_slim.js
./male02/male-02-1noCulling.JPG
./male02/male-02-1noCulling.dds
./male02/male02.mtl
./male02/male02.obj
./male02/male02_dds.mtl
./male02/orig_02_-_Defaul1noCulling.JPG
./male02/orig_02_-_Defaul1noCulling.dds
./male02/readme.txt
./ninja/NinjaLo_bin.bin
./ninja/NinjaLo_bin.js
./terrain.js
./veyron/.htaccess
./veyron/VeyronNoUv_bin.bin
./veyron/VeyronNoUv_bin.js
./veyron/parts/veyron_body_bin.bin
./veyron/parts/veyron_body_bin.js
./veyron/parts/veyron_wheel_bin.bin
./veyron/parts/veyron_wheel_bin.js
./walt/.htaccess
./walt/WaltHead.mtl
./walt/WaltHead.obj

Maybe we could use a (ro)bot to generate this index file with this find or a similar command available for git/github repo each time. And then we would have a javascript function pull only what it needs from the index file to display the full list of available models.

jsidd replied:
Now using this file to load the dropdown list: webgl_loader_viewer.list

looeee commented on examples/webgl_loader_viewer.html:
```diff
@@ -0,0 +1,410 @@
+		// http://stackoverflow.com/questions/11766163/smart-centering-and-scaling-after-model-import-in-three-js
+		// fit camera to object
+		var bBox = new THREE.Box3().setFromObject(scene);
+		var height = bBox.size().y;
```
This gives a warning:

THREE.Box3: .size() has been renamed to .getSize().

jsidd replied:
changed it, thank you.

looeee commented on examples/webgl_loader_viewer.html:
```diff
@@ -0,0 +1,410 @@
+		debugPrint("root children:" + fileScene.children.length+"<br>");
+		var space ="";
+		var childcnt = 0;
+		fileScene.traverse( function ( object ) {
```
Several of the models (.awd, .dae, .gltf, .wrl) throw an error at this point:

three.js:10955 Uncaught TypeError: Cannot read property 'traverse' of undefined(…)

looeee commented on examples/webgl_loader_viewer.html:
```diff
@@ -0,0 +1,410 @@
+			<option value="models/skinned/marine/marine_anims_core.json">models/skinned/marine/marine_anims_core.json</option>
+			<option value="models/assimp/jeep/jeep.assimp.json">models/assimp/jeep/jeep.assimp.json</option>
+			<option value="models/pcd/Zaghetto.pcd">models/pcd/Zaghetto.pcd</option>
+			<option value="models/sea3d/mascot.tjs.sea">models/sea3d/mascot.tjs.sea</option>
```
This model seems to be getting clipped - does the camera far plane need to be greater?

jsidd replied:
I am still working on displaying pcd files properly from the dropdown and textbox. My latest commit added local file selection as well but pcd can not be loaded that way. https://threejs.org/editor does not seem to support pcd at this point either.

looeee replied:
I meant the mascot.tjs.sea file - that one is getting clipped.

jsidd replied:
Right, sorry. I am also looking at it now.

jsidd replied:
fixed it by using an extra param for SEA loader and feeding it the camera name inside the sea model, it has its own node loading mechanism and ignores some settings on standard camera apparently. Also I think I increased the value for the far plane (not yet fully versed in all 3d jargon, sorry) when creating the camera.

looeee commented:
This is great! The only suggestion I'd add would be to extract the loader function into a separate file so that people can use it as a universal loader.

mrdoob commented on examples/webgl_loader_viewer.html:
```diff
@@ -0,0 +1,410 @@
+		<br>
+		Custom model:
+		<input type="text" name="custompath" size="50" id="customModel" value="http://people.sc.fsu.edu/~jburkardt/data/obj/shuttle.obj">
+		<button onclick="reloadModel(getElementById('customModel').value)">Load Custom Model</button>
```
How about using http://threejs.org/editor/ instead? 😁

jsidd replied:
@bchevalier yes, I just saw that with html5 while it is not possible to get a file full path, it is possible to select a local file and get its data.

@mrdoob yes, thank you, I should have look at it earlier; I just tried it, lol :). 

Having a smaller viewer example to just load and view models could be possibly useful? I will study the editor code to see what I can do better.

jsidd replied:
I have added a file picker with my 5th commit.

jsidd committed:
examples/webgl_loader_viewer.html
- fixed/improved node traversal print in debug
- improved adding nodes to scene; however wrl files are now broken (do not display at all)
- added chose file (local asset loader), working formats: babylon, awd, obj, dae files, json (but not .assimp.json),
- current issue with file chooser - texture loading errors (need to figure out how to ignore textures when parsing files
and possibly replace by a white material)

jsidd commented:
> You mention that "the camera currently zooms in on a bounding box but it does not work for all models". Is it because there is an issue with the new THREE.Box3().setFromObject(scene) method?

@bchevalier I still need to look at the camera/bounding box in more details.

mrdoob commented:
The editor already has a similar code:
https://github.com/mrdoob/three.js/blob/dev/editor/js/Loader.js#L26-L423

jsidd committed:
examples/webgl_loader_viewer.html

Moved model viewer js code to examples/js/ModelViewer.js

jsidd commented:
@mrdoob thank you, I looked at it this weekend too. Now reworking my file object based loading code to copy directly these lines. Then I will rework my filepath loading code to work similarly.  

The one difference I see is: it is possible to load some textures directly when using filepath (for ex: dropdown or text field). The file selector functionality however seems to prevent accessing related files (textures...).

jsidd committed:
example/js/ModelViewer.js - now use three.js/editor loading code to load selected local file

also disabled camera changes until I find the right settings

example/webgl_loader_viewer.html - can reselect same file locally to reload the file

jsidd committed:
examples/js/loaders/ModelViewerLoader.js (Loader) - initial THREE.ModelViewerLoader commit

examples/webgl_loader_viewer.html (UI)
- changed local file button,
- replaced reloadModel calls by loadModel
- include new loader js file

examples/js/ModelViewer.js (Controller)
- removed unused functions
- setup new loader uses
- global fileType to handle special fixes

jsidd commented:
Now in three files:
examples/js/loaders/ModelViewerLoader.js (Loader) - initial THREE.ModelViewerLoader commit
examples/webgl_loader_viewer.html (UI)
examples/js/ModelViewer.js (Controller)

jsidd committed:
examples/webgl_loader_viewer.list -> added this index list file with all desired models url

examples/webgl_loader_viewer.html
- now loads dropdown list values from examples/webgl_loader_viewer.list
- moved all javascript code in Controller - /example/js/ModelViewer.js

example/js/ModelViewer.js
- added dropdown list loading from list file function

mrdoob commented on examples/js/ModelViewer.js:
```diff
@@ -0,0 +1,243 @@
+
+ 	var camera, controls, scene, renderer, loader, fileType;
+
+  var fileInput= document.getElementById("selectedFile");
```
Make sure you use tabs for indenting.

jsidd replied:
I just reformatted the indents, is it better now?

jsidd committed:
examples/webgl_loader_viewer.html, examples/js/ModelViewer.js - UI Improvements

jsidd commented:
Now using this file to load the dropdown list: examples/webgl_loader_viewer.list

jsidd committed:
examples/js/ModelViewer.js - reformatted indents

jsidd commented:
I am going to try to move the UI to a collapsible panels using js/libs/dat.gui.min.js 
Then I will look at camera problems and optimizations.

jsidd committed:
examples/js/ModelViewer.js (Controller + new UI)
- use dat.GUI

examples/webgl_loader_viewer.html
- commented out old UI

examples/js/loaders/ModelViewerLoader.js
- fixed handleJSON function to work with model viewer

jsidd commented:
Replaced UI with dat.GUI, fixed some issues with loading JSON from a selected file

jsidd committed:
examples/webgl_loader_viewer.html
- small format fix

examples/js/ModelViewer.js
- changed url initial value to avoid XMLHttpRequest error when not viewing locally

jsidd committed:
- models/sea3d/mascot.tjs.sea working - camera ok
- still working on: pcd (need to flip), duck.gltf (controls and camera stuck), and house.wrl (camera bug)

/example/js/loaders/ModelViewerLoader.js
- set local dae files to load with loader.options.convertUpAxis = true as well
- set sea files to load with custom camera if provided using extra parameter in loadFromPath

/example/js/loaders/ModelViewer.js
- changed default camera :   camera = new THREE.PerspectiveCamera( 45, window.innerWidth / window.innerHeight, 1, 10000 );
- added back filename

jsidd committed:
examples/js/ModelViewer.js
- finally fixed gltf camera, controls issue, apparently this is required: THREE.GLTFLoader.Shaders.update(scene, camera)
as shown below:

function animate() {

  requestAnimationFrame( animate );
  if (fileType === "gltf")
  {
    THREE.GLTFLoader.Shaders.update(scene, camera);
  }
  render();

}

jsidd commented:
Update on recent commits 
- fixed sea model loading (clipping and camera), 
- fixed gltf (duck) loading (apparently `THREE.GLTFLoader.Shaders.update(scene, camera);` is required in `animate()`)

Next: fix wrl camera and probably apply mrdoob coding style guidelines to all my files using the plugin?

jsidd committed:
example/js/ModelViewer.js
- improved pcd camera
- improved wrl camera
- improved code style

example/js/loaders/ModelViewerLoader.js
- improved code style

jsidd commented:
Update : Improved pcd and wrl camera, improved coding style
